### ISSUE TITLE
Prevent violations being sent for resources that are pending deletion

### ISSUE THREAD

**Issue by rich-boyce** [forseti-security/forseti-security#2993]
- Forseti version: 2.17.0
 - Module: notifier or scanner
 - No relevant logs

It seems like Forseti should _not_ send violations for resources that are in the state 'pending_delete'.

Example: A violation is raised for a Cloud SQL instance that is accessible by 0.0.0.0/0. This particular instance was created for testing purposes and is no longer required, so it is deleted. The resource is put into state 'pending_delete', and immediately becomes invisible from the GCP user interfaces, but is still visible to Forseti. Violations continue to be raised by Forseti until the resource is actually removed, approximately a day later. This can cause confusion.

This bug raised as requested by @blueandgold

**Comment by blueandgold** [comment]
Thanks for filing this!  I would think that making the change in the scanner would make sense, as we don't want the violation to be created in the first place.

**Comment by joecheuk** [comment]
Hi @rich-boyce , thanks for reporting the issue, I was doing some investigation but can't seem to find `pending_delete` state  (or any state related to delete) for cloud sql instance.

ref: https://cloud.google.com/sql/docs/mysql/admin-api/v1beta4/instances#resource

state | string | The current serving state of the Cloud SQL instance. This can be one of the following. **RUNNABLE**: The instance is running, or is ready to run when accessed. **SUSPENDED**: The instance is not available, for example due to problems with billing. ### PENDING_CREATE: The instance is being created. `MAINTENANCE`: The instance is down for maintenance. **FAILED**: The instance creation failed. **UNKNOWN_STATE**: The state of the instance is unknown.
-- | -- | --

Do you mind providing more information on where to find the field? We consume data directly from cloud asset inventory API so there is a chance they still have that data cached on their end but without a field clearing stating that the instance is gone, it's hard for us to decide when to remove a field during inventory/scanning time.

**Comment by hshin-g** [comment]
This issue needs more research / verification that the CloudSQL instance being deleted is generating violations.

**Comment by kevensen** [comment]
The field for checking whether a CloudSQL instance is pending deletion is actually found in the the [operations response](https://cloud.google.com/sql/docs/mysql/admin-api/v1beta4/instances/delete#response) after a delete action.

- status: PENDING
- operationType: DELETE

**Comment by blueandgold** [comment]
@kevensen thanks for digging this up!  So, this is a response of the API, and is not something that the Forseti scanner can use.  So, the question is, after a CloudSQL instances is in pending delete state, is Forseti still able to inventory it, and if so, is there anything in the resource representation to indicate the pending delete?

**Comment by kevensen** [comment]
That is correct.  It is a response of the API and not something Forseti can use.  After a quick test of creating and deleting an instance, the instance is not listed when executing `gcloud sql instances list`.

I have no evidence of this at the moment, but my hypothesis is that deleted instance is still showing up in the CAI.  Thus, the Cloud SQL instance is still showing in the inventory table.

**Comment by hshin-g** [comment]
Update: it looks like Cloud SQL instances that are deleted no longer show up in the CAI export data source (and subsequently no violations are raised). Projects that are in "lifecycleState": "DELETE_REQUESTED" are still being ingested, however.

**Comment by gkowalski-google** [comment]
@hshin-g Is this ticket completed? Will we be using [this ticket (#3218)](https://github.com/forseti-security/forseti-security/issues/3218) to track the actual work?

**Comment by hshin-g** [comment]
The one pager has been completed. The one pager only covers pending delete resources and does not cover app-script resources like #3218, so we can either expand the scope of this issue and I can work on the app-script portion in the one pager, or we can just create a new issue or de-scope #3218 for just pending delete resources.
